src/azul/indexer/log_forwarding_service.py

```diff
@@ -0,0 +1,98 @@
+import csv
+import gzip
+from typing import (
+    Iterator,
+)
+
+from botocore.response import (
+    StreamingBody,
+)
+from furl import (
+    furl,
+)
+
+from azul.deployment import (
+    aws,
+)
+from azul.types import (
+    MutableJSON,
+)
+
+
+class LogForwardingService:
+    """
+    Parse logs from an Application Load Balancer.
+    Adapted from https://github.com/rupertbg/aws-load-balancer-logs-to-cloudwatch
+    """
+
+    @property
+    def s3(self):
+        return aws.s3
+
+    def read_logs(self, bucket: str, key: str) -> Iterator[MutableJSON]:
+        response = self.s3.get_object(Bucket=bucket, Key=key)
+        body: StreamingBody = response['Body']
+        with gzip.open(body, mode='rt') as f:
+            for row in csv.reader(f, delimiter=' ', quotechar='"'):
+                fields = dict(zip(self.fields, row))
+                missing = set(self.fields) - fields.keys()
+                assert not missing, missing
+                self.expand_fields(fields)
+                yield fields
+
+    def expand_fields(self, message: MutableJSON) -> None:
+        message.update(self._parse_request(message['request']))
+        for field_name in ['target:port', 'client:port']:
+            message.update(self._parse_port(field_name, message[field_name]))
+
+    def _parse_port(self, field_name: str, element: str) -> MutableJSON:
+        field = field_name.removesuffix(':port')
+        ip, port = element.split(':')
+        return {
+            field: int(ip) if ip == '-1' else ip,
```

Comment:
This deserves an explanation.

Reply:
I don't think it serves any purpose. Just because it was in the example doesn't mean we need to use it.